### ISSUE TITLE
Outdated OX_DYN output

### ISSUE THREAD

**Issue by martijnende** [ydluo/qdyn#42]
The output from `ox_dyn` (in `output.f90::ox_write`) is outdated. The time of the snapshot is incorrect, the values of `dtau_dt` are not initialised.

- [x] Identify the required output quantities (should be the same as for the regular `ox` output
- [x] Create one universal file writer for `ox` and `ox_dyn` output
- [x] Create global `ox_dyn` output unit (currently a magic number `20001+x`)
- [x] Fix non-initialised output

**Comment by jpampuero** [comment]
Two of the three OX_DYN output files can be written re-using the OX procedure. The third file is different, though (rupture time, etc), it requires its own write procedure.
The OX_DYN files are output as one set per snapshot, which is expected by QSB (the QDYN-SPECFEM3D Bridge). The OX outputs have that option too.

**Comment by martijnende** [comment]
What kind of information are the three files supposed to contain? Only `ox` data at the start/end of each rupture?

**Comment by jpampuero** [comment]
File 1 and 2: OX
File 3: info about rupture front, including:
  t_rup = rupture time, the time at which the rupture front passes
  tau_max = peak shear stress 
at each point

**Comment by martijnende** [comment]
While analysing the code line by line, I already found several bugs and issues (MPI gather is called EVERY time step, even when no `ox` output is required!), so `ox_write` definitely needs attention. The main issue is that the business end of the code (writing the output data) is nested 5 levels deep (`if if if if if write endif endif endif endif endif`). As a result, the code that writes the data to disk is duplicated 14 times, each with only a minor tweak to the filename or header information. 

I propose the following strategy to replace all these occurrences:

**During initialisation**

Create arrays of:

1. pointers to the `ox` output quantities (`x`, `v`, `tau`, etc.)
2. `ox` data formats (`3e15.7`, `e24.14`, etc.)
3. pointers to the `ox`, `ox_dyn`, `dyn` header metadata (`it`, `ivmax`, `time`, etc.)

**Each call of `ox_write`**

1. Check which type of `ox` needs to be written
2. Select/update corresponding header pointer array
3. Select output unit (or create a new one in case of `i_ox_seq` etc.)
4. Loop over `mesh%nn` and number of pointers, write to file (one by one)

Relying on arrays of pointers isn't exactly elegant, but at the moment I can't think of anything better (I don't believe F95 supports dictionaries of hash tables?). If you have any other ideas on how to generalise `ox_write` to all the different cases (MPI, `i_ox_seq`, `ox`, `ox_dyn`, etc.), I'd be all ears.

**Comment by jpampuero** [comment]
I think the first step is to fix the indentation and comment each functionality of ox_write (this messy piece of code gives headache!). Then decide which functionalities can be eliminated. We may end up with a small number of cases, for which handling by pointers may not be needed. For instance we can try to:
- simplify the parallel part of the code (if OUT_MASTER ...) by using MPI IO functions that make all processors write to a single file. This may also unify the code for serial and parallel cases.
- stick to binary output, which is more compact (+ modify wrappers accordingly). Consider HDF5.
- output all snapshots in a single file

**Comment by martijnende** [comment]
> I think the first step is to fix the indentation and comment each functionality of ox_write

Done (more or less)

> using MPI IO functions

The idea of MPI IO frightens me...

> unify the code for serial and parallel cases

This would be a good strategy, since serial execution is simply parallel execution on a single core. It would involve some major code refactoring if done properly (separating local and global variables in a more systematic way than is currently done)

> stick to binary output. Consider HDF5

I've pondered over this idea. The main reason for sticking to (optional) ASCII output is long-term accessibility of the data, which is less likely with binary/HDF outputs.

> output all snapshots in a single file

This works (and is default) for regular `ox` output, but not for `ox_dyn` output (one file per event).

**Comment by jpampuero** [comment]
> Done (more or less)

Where?

**Comment by martijnende** [comment]
Here is a translated summary of what's going on:

```
- If parallel
	- If write_ox
		- If OUT_MASTER (default .true.)
			- call pb_global(pb)
			- If MPI_master
				- If i_ox_seq == 0
					- Write ox data sequentially
				- else
					- Create new ox file
					- Write ox data to separate file
					- Close ox file
		- else
			- Write proc meta file
			- Open/create proc data file
			- Write ox data sequentially?
			- Close ox file
	- If i_ox_dyn
		- If OUT_MASTER
			- call pb_global(pb)  =>  should not be called every time step !!!
			- If MPI_master
				- If rising edge
					- Set global parameters
					- Write headers
					- Write ox data to separate file (single snapshot)
					- Close ox file
				- If dynamic
					- If tau > tau_max: Set max stress global parameters
					- If V > V_max: Set max V global parameters
				- If falling edge
					- Write headers
					- Write ox data to separate file (single snapshot)
					- Close ox file

					- Write headers max global values
					- Write max global values
					- Close max global values file
- Else if not parallel
	- If write_ox
		- If i_ox_seq == 0
			- If not BIN_OUTPUT  =>  ox file not closed ??
				- If TP
					- Write header
					- Write ox data
				- Else
					- Write header
					- Write ox data
			- Else
				- Write time?
				- Write some data?
				- If step before stop
					- close ox file
		- Else
			- Open new ox file
			- Write header
			- Write ox data
			- Close ox file
	- If i_ox_dyn
		- If rising edge
			- Set global parameters
			- Write headers
			- Write ox data to separate file (single snapshot)
			- Close ox file
		- If dynamic
			- If tau > tau_max: Set max stress global parameters
			- If V > V_max: Set max V global parameters
		- If falling edge
			- Write headers
			- Write ox data to separate file (single snapshot)
			- Close ox file
			- Write headers max global values
			- Write max global values
			- Close max global values file
	- If DYN_FLAG
		- If rising edge
			- Set global parameters
			- Write headers
			- Write ox data to separate file (single snapshot)
			- Close ox file
		- If falling edge
			- Write headers
			- Write ox data to separate file (single snapshot)
			- Close ox file
```

**Comment by martijnende** [comment]
In most cases, some meta data is written to the output file: `write(unit, '(2i8, e14.6)') pb%it, pb%ot%ivmax, pb%time`. What is the purpose of this? It doesn't seem useful to me in any way.

**Comment by jpampuero** [comment]
A time tag is needed to identify the snapshot. But ivmax seems superfluous.

**Comment by martijnende** [comment]
Time is also written as part of the output (i.e. for each `ox` element), so an additional time tag in the metadata is redundant.

**Comment by jpampuero** [comment]
It takes less disk space if we keep it in the metadata and remove it from the data.

**Comment by martijnende** [comment]
Personally I'd prefer dropping the metadata instead. Removing the time column only saves 10% of space, and breaks backwards compatibility with the wrappers. It'd also be quite a trick to extract the metadata when writing all `ox` data in a single file (right now the Python wrapper simply ignores the metadata).

**Comment by martijnende** [comment]
Currently the Python wrapper only supports ASCII `ox` output collected in a single file. For the time being, I think it makes most sense to only enable/test that type of output in `output.f90` (which would also save me a ton of time testing/debugging other cases). The time series output (`ot`) needs to be revised in a similar way as `ox`, so pruning some output features also saves me some time on that end. Concretely, this implies dropping `BINARY == .true.`, `OUT_MASTER == .false.`, and `i_ox_seq == 1` cases.

For the long term it would be good to consider other output features, like HDF5 or a custom binary file writer.

**Comment by jpampuero** [comment]
These other cases were there for a reason, not necessarily good ones: 
- `BINARY=true` makes much smaller output files and is faster. We implemented it because we needed it. I think it's a useful option to keep.
- `OUT_MASTER=false` was the original version, until we found it cumbersome to deal with a separate file per processor. I think we can now get rid of it.
-  `i_ox_seq=1` was intended to read arbitrary snapshots faster during post-processing, without having to read a whole file. I think there are smarter ways to deal with large files, and we can get rid of this. 

I agree HDF5 would be great. It can also take care of parallel output without complicating the code. In the meantime, our existing custom binary file should do.

**Comment by martijnende** [comment]
The new OX output module seems to work properly for 2D and 3D simulations (including dynamic output). Since the parallel code does not seem to work anymore, none of this has been tested in parallel. While I've included the QDYN-SPECFEM bridge in the OX modules, I've had no way of testing this feature (see #50).

The last item left on the agenda is the binary output: I haven't included binary output in the revisions of the code. We can decide to include HDF support in this release (currently marked as `2.3.0`), or postpone it to later releases (and focus on the implementation of viscosity for now).
